input: convert cap1203 from kscan

faxe1008 opened the pull request:
Convert the CAP1203 driver to the input subsystem, add to build_all tests. 

Some time ago I had a chat with @fabiobaltieri regarding lvgl and the kscan input depedency. Eventually the goal is to create a native lvgl input driver, but before that existing devices using the kscan api need to be converted. Noticing this PR being merged last week [https://github.com/zephyrproject-rtos/zephyr/pull/60161](https://github.com/zephyrproject-rtos/zephyr/pull/60161), I thought I could help making a dent in the ongoing conversion efforts :^)

faxe1008 committed:
input: convert cap1203 from kscan

Convert the CAP1203 driver to the input subsystem, add to build_all tests.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

fabiobaltieri commented on drivers/input/input_cap1203.c:
```diff
@@ -96,7 +95,13 @@ static int cap1203_process(const struct device *dev)
 	}
 
-	data->callback(dev, 0, col, pressed);
+	if (pressed) {
+		input_report_abs(dev, INPUT_ABS_X, col, false, K_FOREVER);
+		input_report_abs(dev, INPUT_ABS_Y, 0, false, K_FOREVER);
+		input_report_key(dev, INPUT_BTN_TOUCH, 1, true, K_FOREVER);
+	} else {
+		input_report_key(dev, INPUT_BTN_TOUCH, 0, true, K_FOREVER);
+	}
```
How would you feel about reworking this one to generate `input_report_key` events instead? This specific driver was somewhat fitting poorly in the keyscan API, so I think that we should convert it properly.

The change would be:
- add a devicetree property to take three input codes and store it in the config structure (you can look at the longpress code for that)
- add a field in the data structure to store the old input state
- in the callback code here compare the new `input` state with the old one, and for every bit that change fire up a `input_report_key` event for the corresponding code and with the right value depending on the new state

That will make it behave the same as `gpio-keys`, down the road we can plumb that in a `LV_INDEV_TYPE_BUTTON` thing.

faxe1008 replied:
I noticed that it was not a kscan driver in the "traditional" sense. If you are ok with breaking downstream, I will perform your suggested changes :^). I'd first commit this separately and if the results are ok, squash it down.

fabiobaltieri replied:
Yeah, I'll check with the architecture folks but I think it's the right thing to do for this.

Fine to do it both in one or two commits, up to you really. Thanks!!

fabiobaltieri reviewed (commented):
Hi Fabian, thanks for the help! Glad to have some of these off my plate. :-) I think this should be converted to generate key events, see my comment on the code. It'd break the downstream application but for this specific case I think it's the right thing to do. Would you do that change?

faxe1008 committed:
input: remove cap1203 kscan-like state report

Previously the driver was retrofitted to the kscan api, handling it as a
input device with one row and three columns. With the move to the input
subsystem each input can have its proper input code instead.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

fabiobaltieri commented on drivers/input/input_cap1203.c:
```diff
@@ -65,24 +69,21 @@ static int cap1203_process(const struct device *dev)
+	for (uint8_t i = 0; i < TOUCH_INPUT_COUNT; i++) {
+		single_input_state = input & BIT(i);
+		if (single_input_state != (data->prev_input_state & BIT(i))) {
+			input_report_key(dev, config->input_codes[i], (single_input_state >> i),
```
No need to shift back the value, `input_report_key` already applies a `!!` to it: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/input/input.h#L82, fine to leave `single_input_state` in here.

fabiobaltieri commented:
Hey @carlescufi, this is a migration from kscan to input for a capsense driver. The x-y-press combination here made no sense though, so we are going to send key events directly instead, like `gpio-keys`. This howere means that the new driver will not work with the kscan compatibility driver and the downstream application will have to be modified in the code.

I'll add a note about it in the release notes, do you think this should also go through https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes? It's not really a change to the API and isolated to a driver so I guess it's down to interpretation. I can put together a RFC for it if necessary.

carlescufi commented:
> Hey @carlescufi, this is a migration from kscan to input for a capsense driver. The x-y-press combination here made no sense though, so we are going to send key events directly instead, like `gpio-keys`. This however means that the new driver will not work with the kscan compatibility driver and the downstream application will have to be modified in the code.
> 
> I'll add a note about it in the release notes, do you think this should also go through https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes? It's not really a change to the API and isolated to a driver so I guess it's down to interpretation. I can put together a RFC for it if necessary.

Looking at this now, I think there is no need for this. This is a bit different than what the stable API change covers: it's not an API that has been modified, but rather an individual driver that has changed the way it exposes its interface to applications. My opinion is that a release note is enough for this.

fabiobaltieri commented:
> Looking at this now, I think there is no need for this. This is a bit different than what the stable API change covers: it's not an API that has been modified, but rather an individual driver that has changed the way it exposes its interface to applications. My opinion is that a release note is enough for this.

Good stuff I agree, thanks for the extra opinion.
